src/freenet/node/diagnostics/threads/DefaultThreadDiagnostics.java

```diff
@@ -0,0 +1,151 @@
+/* This code is part of Freenet. It is distributed under the GNU General
+ * Public License, version 2 (or at your option any later version). See
+ * http://www.gnu.org/ for further details of the GPL. */
+package freenet.node.diagnostics.threads;
+
+import freenet.node.*;
+import freenet.node.diagnostics.*;
+import freenet.support.*;
+
+import java.lang.management.*;
+import java.util.*;
+import java.util.concurrent.atomic.*;
+import java.util.stream.*;
+
+/**
+ * Runnable thread to retrieve node thread's information and compiling it into
+ * an array of NodeThreadInfo objects.
+ */
+public class DefaultThreadDiagnostics implements Runnable, ThreadDiagnostics {
+    private final String name;
+    private final int monitorInterval;
+
+    private final NodeStats nodeStats;
+    private final Ticker ticker;
+
+    /** Sleep interval to calculate % CPU used by each thread */
+    private static final int DEFAULT_MONITOR_INTERVAL = 1000;
+    private static final String DEFAULT_MONITOR_THREAD_NAME = "NodeDiagnostics: thread monitor";
+
+    private final AtomicReference<NodeThreadSnapshot> nodeThreadSnapshot = new AtomicReference<>();
```

Comment:
Totally got thrown off on a tangent here but do we have any nullability annotations somewhere in our dependencies? Because if we had we could generate an empty snapshot here when initializing the AtomicRef and could then annotate the get method below with `@Nonnull`. Just for kicks. (Ignore if we don't have nullability annotations in our dependencies.)

Reply:
I've no clue about this one. I'll check the dependencies though.

Reply:
Apparently there's something like `com.sun.istack.internal`, although it fails to compile. Other than that there's no single dependency pulling the notation.

Reply:
Yes, `com.sun` is generally best avoided and it's also not what I was looking for. Okay, not important. :)

Reply:
Hmm, you know, on the other hand, we could still initialize this here with an empty snapshot which would enable us (haha, you, really 😁) to remove the null check in the render method above. Yes, I think I would prefer that. Null checks are a pest.

Reply:
Great idea! Added the empty snapshot on initialisation. No need for null check and no race conditions can happen now :)